node/core/candidate-selection/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+			)
+			.await
+			{
+				return;
+			}
+
+			let pov = Arc::try_unwrap(pov)
+				.expect("only clone went to fn candidate_is_valid, which is complete; qed");
+
+			match second_candidate(
+				relay_parent,
+				candidate_receipt,
+				pov,
+				&mut self.sender,
+				&self.metrics,
+			)
+			.await
+			{
+				Err(err) => {
+					log::warn!(target: "candidate_selection", "failed to second a candidate: {:?}", err)
+				}
+				Ok(()) => self.seconded_candidate = Some(collator_id),
+			}
+		}
+	}
+
+	async fn handle_invalid(&mut self, candidate_receipt: CandidateReceipt) {
+		let received_from = match &self.seconded_candidate {
+			Some(peer) => peer,
+			None => {
+				log::warn!(target: "candidate_selection", "received invalidity notice for a candidate we don't remember seconding");
+				return;
+			}
+		};
+		log::info!(target: "candidate_selection", "received invalidity note for candidate {:?}", candidate_receipt);
+		let succeeded;
+		if let Err(err) = forward_invalidity_note(received_from, &mut self.sender).await {
+			log::warn!(target: "candidate_selection", "failed to forward invalidity note: {:?}", err);
+			succeeded = false;
+		} else {
+			succeeded = true;
+		}
+		self.metrics.on_invalid_selection(succeeded);
+	}
+}
+
+// get a collation from the Collator Protocol subsystem
+//
+// note that this gets an owned clone of the sender; that's becuase unlike `forward_invalidity_note`, it's expected to take a while longer
+async fn get_collation(
+	relay_parent: Hash,
+	para_id: ParaId,
+	mut sender: mpsc::Sender<FromJob>,
+) -> Result<(CandidateReceipt, PoV), Error> {
+	let (tx, rx) = oneshot::channel();
+	sender
+		.send(FromJob::Collator(CollatorProtocolMessage::FetchCollation(
+			relay_parent,
+			para_id,
+			tx,
+		)))
+		.await?;
+	rx.await.map_err(Into::into)
+}
+
+// find out whether a candidate is valid or not
+async fn candidate_is_valid(
+	candidate_descriptor: CandidateDescriptor,
+	pov: Arc<PoV>,
+	sender: mpsc::Sender<FromJob>,
+) -> bool {
+	std::matches!(
+		candidate_is_valid_inner(candidate_descriptor, pov, sender).await,
+		Ok(true)
+	)
+}
+
+// find out whether a candidate is valid or not, with a worse interface
+// the external interface is worse, but the internal implementation is easier
+async fn candidate_is_valid_inner(
+	candidate_descriptor: CandidateDescriptor,
+	pov: Arc<PoV>,
+	mut sender: mpsc::Sender<FromJob>,
+) -> Result<bool, Error> {
+	let (tx, rx) = oneshot::channel();
+	sender
+		.send(FromJob::Validation(
+			CandidateValidationMessage::ValidateFromChainState(candidate_descriptor, pov, tx),
+		))
+		.await?;
+	Ok(std::matches!(rx.await, Ok(Ok(ValidationResult::Valid(_)))))
+}
+
+async fn second_candidate(
+	relay_parent: Hash,
+	candidate_receipt: CandidateReceipt,
+	pov: PoV,
+	sender: &mut mpsc::Sender<FromJob>,
+	metrics: &Metrics,
+) -> Result<(), Error> {
+	// CandidateBackingMessage::Second(Hash, CandidateReceipt, PoV),
+	match sender
+		.send(FromJob::Backing(CandidateBackingMessage::Second(
```

Comment:
Fairly sure that seconding a candidate just involves sending this message, but if there's more or if the requirements are different, then I'll need to update this function.

Reply:
Yeah, seconding the candidate shouldn't require anything beyond that.